File: optionals/med-4.0.0/tests/c/2.3.6/test24.c
/*  This file is part of MED.
 *
 *  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
 *  MED is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  MED is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with MED.  If not, see <http://www.gnu.org/licenses/>.
 */


/******************************************************************************
 * - Nom du fichier : test24.c
 *
 * - Description : lecture de mailles/faces de type MED_POLYGONE
 *                 dans le maillage MED du fichier test23.med 
 *
 *****************************************************************************/

#include <med.h>
#define MESGERR 1
#include "med_utils.h"
#include <string.h>

#ifdef DEF_LECT_ECR
#define MODE_ACCES MED_LECTURE_ECRITURE
#elif DEF_LECT_AJOUT
#define MODE_ACCES MED_LECTURE_AJOUT
#else
#define MODE_ACCES MED_CREATION
#endif

int main (int argc, char **argv)


{
  med_err ret = 0;
  med_idt fid;
  char maa[MED_TAILLE_NOM+1];  
  med_int nmaa,i,mdim,npoly,j;
  char desc[MED_TAILLE_DESC+1];  
  med_int taille;
  med_int *con, *index, *num, *fam;
  char *nom;
  char tmp[MED_TAILLE_PNOM+1];
  int ind1, ind2,k;
  med_maillage type;

  /* Ouverture du fichier test23.med en lecture seule */
  fid = MEDouvrir("test23.med",MED_LECTURE);
  if (fid < 0) {
    MESSAGE("Erreur a l'ouverture du fichier test23.med");
    return -1;
  }
  printf("Ouverture du fichier test23.med \n");

  /* Lecture du nombre de maillages */
  nmaa = MEDnMaa(fid);
  if (nmaa < 0) {
    MESSAGE("Erreur a la lecture du nombre de maillage");
    return -1;
  }
  printf("Nombre de maillages = "IFORMAT"\n",nmaa);

  for (i=0;i<nmaa;i++) {

    /* Infos sur le maillage */
    if (MEDmaaInfo(fid,i+1,maa,&mdim,&type,desc) < 0) {
      MESSAGE("Erreur a la lecture des infos sur le maillage");
      return -1;
    }
    printf("maillage %d de nom [%s] et de dimension : "IFORMAT" \n",i+1,maa,mdim);
    
    /* Combien de mailles polygones en mode nodal */
    if ((npoly = MEDnEntMaa(fid,maa,MED_CONN,MED_MAILLE,MED_POLYGONE,MED_NOD)) < 0) {
      MESSAGE("Erreur a la lecture du nombre de mailles MED_POLYGONE");
      return -1;
    }
    printf("Nombre de mailles polygones en mode nodal : "IFORMAT" \n",npoly); 

    /* Quelle taille pour le tableau des connectivites */
    if (MEDpolygoneInfo(fid,maa,MED_MAILLE,MED_NOD,&taille) < 0) {
      MESSAGE("Erreur a la lecture des infos sur les maillaes MED_POLYGONE");
      return -1;
    }
    printf("Taille a allouer pour la connectivite des polygones : "IFORMAT" \n",taille); 

    /* Allocation memoire : 
     *  - tableau d'index : npoly + 1
     *  - tableau des connectivites : taille
     *  - tableaux numeros et numeros de familles : npoly
     *  - tableau des noms : MED_TAILLE_PNOM*npoly + 1 
	 */
    index = (med_int *) malloc(sizeof(med_int)*(npoly+1));
    con = (med_int *) malloc(sizeof(med_int)*taille);
    num = (med_int *) malloc(sizeof(med_int)*npoly);
    fam = (med_int *) malloc(sizeof(med_int)*npoly);
    nom = (char *) malloc(sizeof(char)*MED_TAILLE_PNOM*npoly+1);

    /* Lecture de la connectivite des mailles polygones */
    if (MEDpolygoneConnLire(fid,maa,index,npoly+1,con,MED_MAILLE,MED_NOD) < 0) {
      MESSAGE("Erreur a la lecture de la connectivite des mailles MED_POLYGONE");
      ret = -1;
    }
    printf("Lecture de la connectivite des mailles MED_POLYGONE en mode nodal \n");

    /* Lecture noms */
    if (ret == 0) {
      if (MEDnomLire(fid,maa,nom,npoly,MED_MAILLE,MED_POLYGONE) < 0) {
	MESSAGE("Erreur a la lecture des noms des mailles MED_POLYGONE");
	ret = -1;
      }
      printf("Lecture des noms des mailles MED_POLYGONE \n");
    }
    
    /* Lecture des numeros */
    if (ret == 0) {
      if (MEDnumLire(fid,maa,num,npoly,MED_MAILLE,MED_POLYGONE) < 0) {
	MESSAGE("Erreur a la lecture des numeros des mailles MED_POLYGONE");
	ret = -1;
      }
      printf("Lecture des numeros des mailles MED_POLYGONE \n");
    }

    /* lecture des numeros de familles */
    if (ret == 0) {
      if ((ret = MEDfamLire(fid,maa,fam,npoly,MED_MAILLE,MED_POLYGONE)) < 0) {
	MESSAGE("Erreur a la lecture des numeros de famille des mailles MED_POLYGONE");
	ret = -1;
      }
      printf("Lecture des numeros de familles des mailles MED_POLYGONE \n");
    }

    if (ret == 0) {
      printf("Affichage des resultats \n");
      for (j=0;j<npoly;j++) {
	printf(">> Maille MED_POLYGONE "IFORMAT" : \n",j+1);
	printf("---- Connectivite       ----- : [ ");
	ind1 = *(index+j)-1;
	ind2 = *(index+j+1)-1;
	for (k=ind1;k<ind2;k++)
	  printf(IFORMAT" ",*(con+k));
	printf(" ] \n");
	strncpy(tmp,nom+j*MED_TAILLE_PNOM,MED_TAILLE_PNOM);
	tmp[MED_TAILLE_PNOM] = '\0';
	printf("---- Nom                ----- : %s \n",tmp);
	printf("---- Numero             ----- : "IFORMAT" \n",*(num+j));
	printf("---- Numero de famille  ----- : "IFORMAT" \n",*(fam+j));
      }
    }

    /* Liberation de la memoire */
    free(index);
    free(con);
    free(num);
    free(fam);
    free(nom);
  }

  /* Fermeture du fichier */
  if (MEDfermer(fid) < 0) {
    MESSAGE("Erreur a la fermeture du fichier");
    return -1;
  }
  printf("Fermeture du fichier \n");
 
  return ret; 
}

File: optionals/med-4.0.0/tests/f/2.3.6/test24.f
C*  This file is part of MED.
C*
C*  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
C*  MED is free software: you can redistribute it and/or modify
C*  it under the terms of the GNU Lesser General Public License as published by
C*  the Free Software Foundation, either version 3 of the License, or
C*  (at your option) any later version.
C*
C*  MED is distributed in the hope that it will be useful,
C*  but WITHOUT ANY WARRANTY; without even the implied warranty of
C*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
C*  GNU Lesser General Public License for more details.
C*
C*  You should have received a copy of the GNU Lesser General Public License
C*  along with MED.  If not, see <http://www.gnu.org/licenses/>.
C*

C       *******************************************************************************
C       * - Nom du fichier : test24.f
C       *
C       * - Description : lecture de mailles MED_POLYGONE dans le maillage MED
C       *                 du fichier test23.med   
C       *
C       ******************************************************************************
	program test23
C       
	implicit none
	include 'med.hf'
C	
	integer*8 fid
	integer cret,mdim,nmaa,npoly,i,j,k,taille
	character*32 maa
 	character*200 desc
        integer ni, n
	parameter (ni=4, n=3)
	integer index(ni),ind1,ind2
	character*16 nom(n)
	integer num(n),fam(n)
        integer con(16)
	integer type
C
C       Ouverture du fichier test23.med en lecture seule
 	call efouvr(fid,'test23.med',MED_LECTURE, cret)
        print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur ouverture du fichier'
	   call efexit(-1)
	endif      
        print *,'Ouverture du fichier test23.med'
C
C       Lecture du nombre de maillages
	call efnmaa(fid,nmaa,cret)
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur lecture nombre de maillage'
	   call efexit(-1)
	endif      
	print *,'Nombre de maillages : ',nmaa
C   
C       Lecture de toutes les mailles MED_POLYGONE
C       dans chaque maillage
	do 10 i=1,nmaa
C
C          Info sur chaque maillage
	   call efmaai(fid,i,maa,mdim,type,desc,cret)
	   if (cret .ne. 0 ) then
	      print *,'Erreur lecture infos maillage'
	      call efexit(-1)
	   endif      
	   print *,cret
	   print *,'Maillage : ',maa
	   print *,'Dimension : ',mdim
C     
C          Combien de mailles polygones
	   call efnema(fid,maa,MED_CONN,MED_MAILLE,MED_POLYGONE,
     &                       MED_NOD,npoly,cret)
	   print *,cret
	   if (cret .ne. 0 ) then
	      print *,'Erreur lecture du nombre de polygone'
	      call efexit(-1)
	   endif      
	   print *,'Nombre de mailles MED_POLYGONE : ',npoly
C
C          Taille des connectivites
	   call efpygi(fid,maa,MED_MAILLE,MED_NOD,taille,cret)
	   print *,cret
	   if (cret .ne. 0 ) then
	      print *,'Erreur lecture infos polygones'
	      call efexit(-1)
	   endif      
	   print *,'Taille de la connectivite : ',taille
C
C          Lecture de la connectivite
	   call efpgcl(fid,maa,index,npoly+1,con,MED_MAILLE,
     &                       MED_NOD,cret)
	   print *,cret
	   if (cret .ne. 0 ) then
	      print *,'Erreur lecture des connectivites polygones'
	      call efexit(-1)
	   endif      
	   print *,'Lecture de la connectivite des polygones'
C
C          Lecture des noms
	   call efnoml(fid,maa,nom,npoly,MED_MAILLE,MED_POLYGONE,
     &                       cret)
	   print *,cret
	   if (cret .ne. 0 ) then
	      print *,'Erreur lecture des noms des polygones'
	      call efexit(-1)
	   endif      
	   print *,'Lecture des noms'
C
C          Lecture des numeros
	   call efnuml(fid,maa,num,npoly,MED_MAILLE,MED_POLYGONE,
     &                       cret)
	   print *,cret
	   if (cret .ne. 0 ) then
	      print *,'Erreur lecture des numeros des polygones'
	      call efexit(-1)
	   endif      
	   print *,'Lecture des numeros'
C
C          Lecture des numeros de familles
	   call effaml(fid,maa,fam,npoly,MED_MAILLE,MED_POLYGONE,
     &                       cret)
	   print *,cret
	   if (cret .ne. 0 ) then
	      print *,'Erreur lecture des numeros de famille des
     & polygones'
	      call efexit(-1)
	   endif      
	   print *,'Lecture des numeros de famille'
C
C          Affichage des resultats
	   print *,'Affichage des resultats'
	   do 20 j=1,npoly
C       
	      print *,'>> Maille polygone ',j
	      print *,'---- Connectivite      ---- : '
	      ind1 = index(j)
	      ind2 = index(j+1)
	      do 30 k=ind1,ind2-1
		 print *,con(k)
 30	      continue
	      print *,'---- Nom               ---- : ',nom(j)
	      print *,'---- Numero            ----:  ',num(j)
	      print *,'---- Numero de famille ---- : ',fam(j)
C
 20	   continue
C
 10	continue
C
C       Fermeture du fichier
	call efferm (fid,cret)
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur fermeture du fichier'
	   call efexit(-1)
	endif      
	print *,'Fermeture du fichier'
C
        end
